x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java

```diff
@@ -122,37 +141,66 @@ private void handleFailure(Exception e, ActionListener<GetPipelineResponse> list
         }
     }
 
-    private void handleSearchResponse(
+    private void handleFilteringSearchResponse(
         SearchResponse searchResponse,
         Map<String, BytesReference> pipelineSources,
+        Set<String> explicitPipelineIds,
+        Set<String> wildcardPipelineIdPatterns,
+        int maxNumHitsToFetch,
         Consumer<SearchResponse> clearScroll,
         ActionListener<GetPipelineResponse> listener
     ) {
+        maxNumHitsToFetch += searchResponse.getHits().getHits().length;
         for (SearchHit hit : searchResponse.getHits().getHits()) {
-            pipelineSources.put(hit.getId(), hit.getSourceRef());
+            if (explicitPipelineIds.isEmpty() && wildcardPipelineIdPatterns.isEmpty()) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+
+            // take if ID is in request IDs set
+            if (explicitPipelineIds.contains(hit.getId())) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+            // take if id matches request wildcard pattern
+            for (String wildcardPipelineIdPattern : wildcardPipelineIdPatterns) {
+                if (Pattern.matches(wildcardPipelineIdPattern, hit.getId())) {
+                    pipelineSources.put(hit.getId(), hit.getSourceRef());
+                    break;
+                }
+            }
         }
 
-        if (pipelineSources.size() > searchResponse.getHits().getTotalHits().value) {
+        if (maxNumHitsToFetch > searchResponse.getHits().getTotalHits().value) {
```

Comment:
I think we can check for this failure scenario before iterating over the records from this scroll response. I think it'd be better to error out earlier in that case and save the work of checking against patterns.